Bottom: 821d9efd730bdf2dce61c9c049639a5a8de226d8
Top:    de3507c92aa28779deceb41ef5977a540d91c474
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-12-04 15:56:13 +0000

Fix timeout forced redraw to run from the main loop, not directly

This avoids us drawing when the board is in an intermediate state. Will
likely not give as much immediacy as drawing from where the graphics are
invalidated - but avoids visual artaefacts when drawing tracks with
auto-enforce DRC enabled.


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 8229e63..8e3b9f4 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -925,6 +925,14 @@ ghid_invalidate_lr (int left, int right, int top, int bottom)
   ghid_invalidate_all ();
 }
 
+static gboolean
+force_redraw (gpointer user_data)
+{
+  gdk_window_process_all_updates ();
+
+  return G_SOURCE_REMOVE;
+}
+
 #define MAX_ELAPSED (50. / 1000.) /* 50ms */
 void
 ghid_invalidate_all ()
@@ -935,7 +943,7 @@ ghid_invalidate_all ()
   ghid_draw_area_update (gport, NULL);
 
   if (elapsed > MAX_ELAPSED)
-    gdk_window_process_all_updates ();
+    g_idle_add_full (G_PRIORITY_HIGH, force_redraw, NULL, NULL);
 }
 
 void
